Cleanup unused Google Ads services

mikkamp opened the pull request:
### Changes proposed in this Pull Request:

The main purpose of this PR is to reduce the size of the Google Ads library. The large size of the library, and the many generated files cause problems on some hosts that timeout while installing/updating the plugin.
Since we only use a small part of the library this PR creates a cleanup script which removes services and enums from the library which we don't use.

Running the script reduces the size in the following way:
- Final zip file: 10.7MB > 7MB = 3.7MB smaller
- Files in zip file: 33.6MB > 22.2MB = 11.4MB smaller
- Googleads directory: 18.3MB > 8.2MB = 10.1MB smaller

![image](https://user-images.githubusercontent.com/11388669/167866430-b3b6e5c8-d434-4d6f-9d0e-642f748d6e11.png)

Resolves part of #314 

### Screenshots:
![image](https://user-images.githubusercontent.com/11388669/167582528-829ac443-ca96-4bc8-8351-b13ec491b224.png)

### Detailed test instructions:
1. Run composer install (which will run the cleanup script) 
2. Run the unit tests to confirm that all the needed files from the Ads library are still available
3. Run `npm run build`
4. Check the final zip file and confirm it's reduced in size

### Additional details:
The cleanup script is directly tied to the library version. If we update the library version the cleanup script will need to be modified and any newly added services will need to be added to the list. The list of services was taken from vendor/googleads/google-ads-php/src/Google/Ads/GoogleAds/Lib/V9/ServiceClientFactoryTrait.php

### Changelog entry
* Tweak - Cleanup unused Google Ads services.

mikkamp committed:
Initial helper script

mikkamp committed:
Remove metadata files

mikkamp committed:
Add debug output

mikkamp committed:
Reset source path

mikkamp committed:
Remove additional files

mikkamp committed:
Cleanup services in class override

mikkamp committed:
Amend services list

mikkamp committed:
Output library version we are replacing

mikkamp committed:
Run Ads cleanup script from composer

mikkamp committed:
Remove additional resources

mikkamp committed:
Find and remove unused enums

puntope commented on bin/GoogleAdsCleanupServices.php:
```diff
@@ -0,0 +1,529 @@
+	 *
+	 * @var array
+	 */
+	protected $services_keep = [
+		'AccountBudget',
+		'AccountLink',
+		'Ad',
+		'AdGroup',
+		'AdGroupAdLabel',
+		'AdGroupAd',
+		'AdGroupCriterion',
+		'AssetGroup',
+		'AssetGroupListingGroupFilter',
+		'BillingSetup',
+		'Campaign',
+		'CampaignBudget',
+		'CampaignCriterion',
+		'ConversionAction',
+		'Customer',
+		'CustomerUserAccess',
+		'GeoTargetConstant',
+		'GoogleAds',
+		'MerchantCenterLink',
+	];
```
💅  This is not being used anywhere

mikkamp replied:
It was mainly left in to know which services are being used when having to update to a newer version of the library. But if we refactor that, then this list can be removed as well as the list of services to remove.

puntope commented on bin/GoogleAdsCleanupServices.php:
```diff
@@ -0,0 +1,529 @@
+	 *
+	 * @var array
+	 */
+	protected $services = [
+		'AccountBudgetProposal',
+		'AdGroupAdAssetView',
+		'AdGroupAudienceView',
+		'AdGroupBidModifier',
+		'AdGroupCriterionCustomizer',
+		'AdGroupCriterionLabel',
+		'AdGroupCriterionSimulation',
+		'AdGroupCustomizer',
+		'AdGroupExtensionSetting',
+		'AdGroupFeed',
+		'AdGroupLabel',
+		'AdGroupSimulation',
+		'AdParameter',
+		'AdScheduleView',
+		'AgeRangeView',
+		'Asset',
+		'AssetFieldTypeView',
+		'AssetGroupProductGroupView',
+		'AssetSet',
+		'BatchJob',
+		'BiddingDataExclusion',
+		'BiddingSeasonalityAdjustment',
+		'BiddingStrategy',
+		'BiddingStrategySimulation',
+		'CallView',
+		'CampaignAsset',
+		'CampaignAssetSet',
```
❓ 
I'm wondering if instead of this manual list we can just read the file `src/Google/Ads/ServiceClientFactoryTrait.php`  and grab the used services there (as well as the version) 

Then, delete all the Services in the Services folder which are not being used. 

Kinda the same as we did for Enum.

mikkamp replied:
That's a good point. The code for removing services was written a while back when we didn't have full testing in place, so I didn't want a missed import to cause the service to be removed from the library. Now that things have matured it might be better to implement a similar method as for the Enums. It will definitely make it easier to maintain, in case we update the library to a newer version.

puntope commented on bin/GoogleAdsCleanupServices.php:
```diff
@@ -0,0 +1,529 @@
+		$this->remove_use_statement( $file, "{$service}ServiceClient" );
+		$this->remove_function( $file, "get{$service}ServiceClient" );
+
+		$this->remove_file( "/{$this->version}/Services/Gapic/{$service}ServiceGapicClient.php" );
```
💅 

Maybe I missed some of them but I guess `this->version` is added in every `remove_file` function call. HEnce I guess we can just add that part of the path directly in there.

mikkamp replied:
I decided not to implement this change, while we technically can move the version to the `remove_file` function it would make the function inconsistent with the behaviour of `remove_function` / `remove_use_statement`.

Based on the folder structure we can sometimes have `Lib/<version>...` and other times `<version>/Enums...` so I prefer to be consistent in the file string we pass to the functions.

puntope commented on bin/GoogleAdsCleanupServices.php:
```diff
@@ -0,0 +1,529 @@
+
+		return array_map(
+			function( $file ) {
+				preg_match( '/(.*)\.php/', basename( $file ), $matches );
+				return $matches[1];
```
💅 

`return pathinfo($file, PATHINFO_FILENAME);` also works and is easier to read.

puntope commented on bin/GoogleAdsCleanupServices.php:
```diff
@@ -0,0 +1,529 @@
+	 * @return array
+	 */
+	protected function find_used_enums(): array {
+		$pattern = "use Google\\\\Ads\\\\GoogleAds\\\\{$this->version}\\\\Enums\\\\([A-Za-z]+)Enum\\\\";
```
❓ 

With `([A-Za-z]+)` we cover letters only. What if the Enum name has a number?

mikkamp replied:
They don't use any names with a number, but I've added it anyways just in case.

puntope commented on bin/GoogleAdsCleanupServices.php:
```diff
@@ -0,0 +1,529 @@
+	 * @param string $file
+	 * @param string $name Function name.
+	 */
+	protected function remove_function( string $file, string $name ) {
+		$file = $this->file_path( $file );
+		if ( ! file_exists( $file ) ) {
+			$this->warning( sprintf( 'File does not exist: %s', $file ) );
+			return;
+		}
+
+		$contents = file_get_contents( $file ); // phpcs:ignore WordPress.WP.AlternativeFunctions
```
❓ 

It's necessary to remove those 2 functions?

mikkamp replied:
Do you mean the function from ResourceNames and ServiceClientFactoryTrait? It's not strictly necessary but it's cleaner then leaving a bunch of functions around that can't be called or they would result in an error. We also override ServiceClientFactoryTrait which is easier to override/update if we have a smaller set of functions.

puntope replied:
So my worries are about how safe is this function  and how potentially can create problems parsing and deleting text or code we don't control that could change over time. 

For example introducing a bracket in a comment like inside the function body (just a random example).

Since post install script is not failing for using an undefined class inside a never invoked function and since this script is for reducing space rather than cleaning the Google Code I was wondering if its worth it to remove those functions.

But was just an observation and it's not a blocker.

mikkamp replied:
I agree that those functions don't cover us 100% and there are flaws. However if I look at the bigger picture, I do think it's still worth it to continue removing them.

Initially we asked the owners of the library for a script to remove unused services, there was desire to implement it but it didn't have a high priority. So this script is our interim solution to slim down the library solely for our use case. It never set out to be a complete solution.

Even though we don't control the code, updating the library is a manual process which is done through a PR because we override classes to make use of the library. So if any part of this script would break a function for future versions of the library, it would be caught when running the unit tests in the PR and it won't go unnoticed.

puntope commented on bin/GoogleAdsCleanupServices.php:
```diff
@@ -0,0 +1,529 @@
+	 */
+	protected function remove_file( string $file ) {
+		$file = $this->file_path( $file );
+		if ( ! file_exists( $file ) ) {
+			$this->warning( sprintf( 'File does not exist: %s', $file ) );
+			return;
+		}
```
💅 

Duplicated code in 445

mikkamp replied:
I personally think since this is a helper script and not included in the final code. It's an ok balance to have this duplicated code instead of abstracting everything into separate functions. Since we aren't using exception handling even if a helper function would be used it would still need to check a condition to return early, so there isn't a lot of duplication you can prevent here.

puntope commented on src/Google/Ads/ServiceClientFactoryTrait.php:
```diff
@@ -13,158 +13,25 @@
 
 use Google\Ads\GoogleAds\Constants;
 use Google\Ads\GoogleAds\Lib\ConfigurationTrait;
-use Google\Ads\GoogleAds\V9\Services\AccessibleBiddingStrategyServiceClient;
-use Google\Ads\GoogleAds\V9\Services\AccountBudgetProposalServiceClient;
 use Google\Ads\GoogleAds\V9\Services\AccountBudgetServiceClient;
 use Google\Ads\GoogleAds\V9\Services\AccountLinkServiceClient;
```
❔ 

This is just for my personal curiosity, how we know which of these files and dependencies we should remove?

mikkamp replied:
The list was manually created by checking all the Services in the original file and comparing it with any of the ones we used within our code. Now that we switched to automatically detecting which ones we use, this would no longer be needed, which makes the script easier to maintain when switching library versions.

puntope reviewed (commented):
Howdy @mikkamp 

I tested this and overall  it's looking great. A huge improvement in the plugin size 👏 

I left some comments and questions to be discussed.

I'm  now performing a Smoke test just to be sure nothing is going wrong.

mikkamp committed:
Use pathinfo to get filename without extension

mikkamp committed:
Search for list of used services

mikkamp committed:
Remove list of services

mikkamp committed:
Merge develop

mikkamp commented:
@puntope Thanks for the review. I've reworked the code to search for the list of Services/Resources in the same way we searched for Enums. I've also addressed the points you mentioned, some I added a comment as I chose not to implement it. It's ready for another round of review.

mikkamp committed:
Fix remove suffix

puntope commented on bin/GoogleAdsCleanupServices.php:
```diff
@@ -233,12 +233,34 @@ public static function remove( Event $event = null ) {
 
 	/**
-	 * Remove all services.
+	 * Remove all services and resources.
```
💅  Super minor, but being strict this should be `Remove all unused services and resources.`

puntope commented on bin/GoogleAdsCleanupServices.php:
```diff
@@ -0,0 +1,410 @@
+	 * @return string
+	 */
+	protected function remove_suffix( string $suffix, string $text ): string {
+		return preg_replace( '/' . preg_quote( $suffix, '/' ) . '$/', '', $text );
```
❓ 

Why do we need to escape  '/' in the `preg_quote`?  As far as I know `pathinfo( $file, PATHINFO_FILENAME )` will return just the file name with no `/` cases.

mikkamp replied:
We aren't just escaping `/`, we are passing that as the delimiter character. We are escaping any [special character](https://www.php.net/manual/en/function.preg-quote.php).

Not strictly necessary for our use case since the suffix contains only letters. But since it's a function called `remove_suffix` it could be used for other strings. And you could also argue that `PATHINFO_FILENAME` could in some cases return a string with a `.` which is on the list of special characters.

puntope commented on bin/GoogleAdsCleanupServices.php:
```diff
@@ -0,0 +1,410 @@
+	 * Find a specific filename pattern within the library.
+	 *
+	 * @param string $pattern Regexp pattern to match.
+	 * @param string $suffix  Suffix to remove from filename.
```
```suggestion
	 * @param string|null $suffix  Suffix to remove from filename.
```

puntope reviewed (approved):
LGTM ✅ 

@mikkamp Looking good. Now this util will be maintainable over the time easier. 🥳 

I tested the post-install and works ✅ 
I tested to create a new `MyTestService.php` in the metadata as well as `MyTestServiceClient.php` in the Services folder and it was deleted as expected. ✅ 
Smoked test and unit tests worked for me as well ✅ 

Files deleted in my test: 

<details>
<summary>Click to see the list of files removed in my test</summary>
Removing unused services and resources from Google Ads library V9
Removing service AccessibleBiddingStrategy
Removing service AccountBudgetProposal
Removing service AdGroupAdAssetView
Removing service AdGroupAsset
Removing service AdGroupAudienceView
Removing service AdGroupBidModifier
Removing service AdGroupCriterionCustomizer
Removing service AdGroupCriterionLabel
Removing service AdGroupCriterionSimulation
Removing service AdGroupCustomizer
Removing service AdGroupExtensionSetting
Removing service AdGroupFeed
Removing service AdGroupLabel
Removing service AdGroupSimulation
Removing service AdParameter
Removing service AdScheduleView
Removing service AgeRangeView
Removing service AssetFieldTypeView
Removing service AssetGroupAsset
Removing service Asset
Removing service AssetSetAsset
Removing service AssetSet
Removing service BatchJob
Removing service BiddingDataExclusion
Removing service BiddingSeasonalityAdjustment
Removing service BiddingStrategy
Removing service BiddingStrategySimulation
Removing service CampaignAsset
Removing service CampaignAssetSet
Removing service CampaignAudienceView
Removing service CampaignBidModifier
Removing service CampaignConversionGoal
Removing service CampaignCriterionSimulation
Removing service CampaignCustomizer
Removing service CampaignDraft
Removing service CampaignExperiment
Removing service CampaignExtensionSetting
Removing service CampaignFeed
Removing service CampaignLabel
Removing service CampaignSharedSet
Removing service CampaignSimulation
Removing service CarrierConstant
Removing service ChangeStatus
Removing service ClickView
Removing service CombinedAudience
Removing service ConversionAdjustmentUpload
Removing service ConversionCustomVariable
Removing service ConversionGoalCampaignConfig
Removing service ConversionUpload
Removing service ConversionValueRule
Removing service ConversionValueRuleSet
Removing service CurrencyConstant
Removing service CustomAudience
Removing service CustomConversionGoal
Removing service CustomInterest
Removing service CustomerAsset
Removing service CustomerClientLink
Removing service CustomerClient
Removing service CustomerConversionGoal
Removing service CustomerCustomizer
Removing service CustomerExtensionSetting
Removing service CustomerFeed
Removing service CustomerLabel
Removing service CustomerManagerLink
Removing service CustomerNegativeCriterion
Removing service CustomerUserAccessInvitation
Removing service CustomizerAttribute
Removing service DetailPlacementView
Removing service DetailedDemographic
Removing service DisplayKeywordView
Removing service DistanceView
Removing service DomainCategory
Removing service DynamicSearchAdsSearchTermView
Removing service ExpandedLandingPageView
Removing service ExtensionFeedItem
Removing service FeedItem
Removing service FeedItemSetLink
Removing service FeedItemSet
Removing service FeedItemTarget
Removing service FeedMapping
Removing service FeedPlaceholderView
Removing service Feed
Removing service GenderView
Removing service GeographicView
Removing service GoogleAdsField
Removing service GroupPlacementView
Removing service HotelGroupView
Removing service HotelPerformanceView
Removing service IncomeRangeView
Removing service Invoice
Removing service KeywordPlanAdGroupKeyword
Removing service KeywordPlanAdGroup
Removing service KeywordPlanCampaignKeyword
Removing service KeywordPlanCampaign
Removing service KeywordPlanIdea
Removing service KeywordPlan
Removing service KeywordThemeConstant
Removing service KeywordView
Removing service Label
Removing service LandingPageView
Removing service LanguageConstant
Removing service LifeEvent
Removing service LocationView
Removing service ManagedPlacementView
Removing service MediaFile
Removing service MobileAppCategoryConstant
Removing service MobileDeviceConstant
Removing service OfflineUserDataJob
Removing service OperatingSystemVersionConstant
Removing service PaidOrganicSearchTermView
Removing service ParentalStatusView
Removing service PaymentsAccount
Removing service ProductBiddingCategoryConstant
Removing service ProductGroupView
Removing service ReachPlan
Removing service Recommendation
Removing service RemarketingAction
Removing service SearchTermView
Removing service SharedCriterion
Removing service SharedSet
Removing service ShoppingPerformanceView
Removing service SmartCampaignSearchTermView
Removing service SmartCampaignSetting
Removing service SmartCampaignSuggest
Removing service ThirdPartyAppAnalyticsLink
Removing service TopicConstant
Removing service TopicView
Removing service UserData
Removing service UserInterest
Removing service UserList
Removing service UserLocationView
Removing service Video
Removing service WebpageView
Removing service AssetGroupProductGroupView
Removing service CallView
Removing service ChangeEvent
Removing service HotelReconciliation
Removing unused enums from Google Ads library V9
Removing enum AccessInvitationStatus
Removing enum AccessReason
Removing enum AccountBudgetProposalStatus
Removing enum AccountBudgetProposalType
Removing enum AccountBudgetStatus
Removing enum AccountLinkStatus
Removing enum AdCustomizerPlaceholderField
Removing enum AdDestinationType
Removing enum AdGroupAdRotationMode
Removing enum AdGroupAdStatus
Removing enum AdGroupCriterionApprovalStatus
Removing enum AdGroupCriterionStatus
Removing enum AdGroupStatus
Removing enum AdGroupType
Removing enum AdNetworkType
Removing enum AdServingOptimizationStatus
Removing enum AdStrength
Removing enum AdType
Removing enum AdvertisingChannelSubType
Removing enum AffiliateLocationFeedRelationshipType
Removing enum AffiliateLocationPlaceholderField
Removing enum AgeRangeType
Removing enum AppCampaignAppStore
Removing enum AppCampaignBiddingStrategyGoalType
Removing enum AppPaymentModelType
Removing enum AppPlaceholderField
Removing enum AppStore
Removing enum AppUrlOperatingSystemType
Removing enum AssetFieldType
Removing enum AssetLinkStatus
Removing enum AssetPerformanceLabel
Removing enum AssetSetAssetStatus
Removing enum AssetSetLinkStatus
Removing enum AssetSetStatus
Removing enum AssetSetType
Removing enum AssetType
Removing enum AttributionModel
Removing enum BatchJobStatus
Removing enum BidModifierSource
Removing enum BiddingSource
Removing enum BiddingStrategyStatus
Removing enum BiddingStrategyType
Removing enum BrandSafetySuitability
Removing enum BudgetCampaignAssociationStatus
Removing enum BudgetDeliveryMethod
Removing enum BudgetPeriod
Removing enum BudgetStatus
Removing enum BudgetType
Removing enum CallConversionReportingState
Removing enum CallPlaceholderField
Removing enum CallToActionType
Removing enum CallTrackingDisplayLocation
Removing enum CallType
Removing enum CalloutPlaceholderField
Removing enum CampaignDraftStatus
Removing enum CampaignExperimentStatus
Removing enum CampaignExperimentTrafficSplitType
Removing enum CampaignExperimentType
Removing enum CampaignServingStatus
Removing enum CampaignSharedSetStatus
Removing enum ChangeClientType
Removing enum ChangeEventResourceType
Removing enum ChangeStatusOperation
Removing enum ChangeStatusResourceType
Removing enum ClickType
Removing enum CombinedAudienceStatus
Removing enum ContentLabelType
Removing enum ConversionActionCountingType
Removing enum ConversionAdjustmentType
Removing enum ConversionAttributionEventType
Removing enum ConversionCustomVariableStatus
Removing enum ConversionLagBucket
Removing enum ConversionOrAdjustmentLagBucket
Removing enum ConversionOrigin
Removing enum ConversionValueRulePrimaryDimension
Removing enum ConversionValueRuleSetStatus
Removing enum ConversionValueRuleStatus
Removing enum CriterionCategoryChannelAvailabilityMode
Removing enum CriterionCategoryLocaleAvailabilityMode
Removing enum CriterionSystemServingStatus
Removing enum CriterionType
Removing enum CustomAudienceMemberType
Removing enum CustomAudienceStatus
Removing enum CustomAudienceType
Removing enum CustomConversionGoalStatus
Removing enum CustomInterestMemberType
Removing enum CustomInterestStatus
Removing enum CustomInterestType
Removing enum CustomPlaceholderField
Removing enum CustomerMatchUploadKeyType
Removing enum CustomerPayPerConversionEligibilityFailureReason
Removing enum CustomizerAttributeStatus
Removing enum CustomizerAttributeType
Removing enum CustomizerValueStatus
Removing enum DataDrivenModelStatus
Removing enum DayOfWeek
Removing enum Device
Removing enum DisplayAdFormatSetting
Removing enum DisplayUploadProductType
Removing enum DistanceBucket
Removing enum DsaPageFeedCriterionField
Removing enum EducationPlaceholderField
Removing enum ExtensionSettingDevice
Removing enum ExtensionType
Removing enum ExternalConversionSource
Removing enum FeedAttributeType
Removing enum FeedItemQualityApprovalStatus
Removing enum FeedItemQualityDisapprovalReason
Removing enum FeedItemSetStatus
Removing enum FeedItemSetStringFilterType
Removing enum FeedItemStatus
Removing enum FeedItemTargetDevice
Removing enum FeedItemTargetStatus
Removing enum FeedItemTargetType
Removing enum FeedItemValidationStatus
Removing enum FeedLinkStatus
Removing enum FeedMappingCriterionType
Removing enum FeedMappingStatus
Removing enum FeedOrigin
Removing enum FeedStatus
Removing enum FlightPlaceholderField
Removing enum FrequencyCapEventType
Removing enum FrequencyCapLevel
Removing enum FrequencyCapTimeUnit
Removing enum GenderType
Removing enum GeoTargetConstantStatus
Removing enum GeoTargetingRestriction
Removing enum GeoTargetingType
Removing enum GoalConfigLevel
Removing enum GoogleAdsFieldCategory
Removing enum GoogleAdsFieldDataType
Removing enum GoogleVoiceCallStatus
Removing enum HotelDateSelectionType
Removing enum HotelPlaceholderField
Removing enum HotelPriceBucket
Removing enum HotelRateType
Removing enum HotelReconciliationStatus
Removing enum ImagePlaceholderField
Removing enum IncomeRangeType
Removing enum InteractionEventType
Removing enum InteractionType
Removing enum InvoiceType
Removing enum JobPlaceholderField
Removing enum KeywordMatchType
Removing enum KeywordPlanAggregateMetricType
Removing enum KeywordPlanCompetitionLevel
Removing enum KeywordPlanConceptGroupType
Removing enum KeywordPlanForecastInterval
Removing enum KeywordPlanKeywordAnnotation
Removing enum KeywordPlanNetwork
Removing enum LabelStatus
Removing enum LeadFormCallToActionType
Removing enum LeadFormDesiredIntent
Removing enum LeadFormFieldUserInputType
Removing enum LeadFormPostSubmitCallToActionType
Removing enum LegacyAppInstallAdAppStore
Removing enum LinkedAccountType
Removing enum ListingGroupFilterBiddingCategoryLevel
Removing enum ListingGroupFilterCustomAttributeIndex
Removing enum ListingGroupFilterProductChannel
Removing enum ListingGroupFilterProductCondition
Removing enum ListingGroupFilterProductTypeLevel
Removing enum ListingGroupFilterTypeEnum
Removing enum ListingGroupType
Removing enum LocalPlaceholderField
Removing enum LocationExtensionTargetingCriterionField
Removing enum LocationGroupRadiusUnits
Removing enum LocationPlaceholderField
Removing enum LocationSourceType
Removing enum ManagerLinkStatus
Removing enum MatchingFunctionContextType
Removing enum MatchingFunctionOperator
Removing enum MediaType
Removing enum MessagePlaceholderField
Removing enum MimeType
Removing enum MinuteOfHour
Removing enum MobileAppVendor
Removing enum MobileDeviceType
Removing enum MonthOfYear
Removing enum NegativeGeoTargetType
Removing enum OfflineUserDataJobFailureReason
Removing enum OfflineUserDataJobMatchRateRange
Removing enum OfflineUserDataJobStatus
Removing enum OfflineUserDataJobType
Removing enum OperatingSystemVersionOperatorType
Removing enum OptimizationGoalType
Removing enum ParentalStatusType
Removing enum PaymentMode
Removing enum PlaceholderType
Removing enum PlacementType
Removing enum PolicyApprovalStatus
Removing enum PolicyReviewStatus
Removing enum PolicyTopicEntryType
Removing enum PolicyTopicEvidenceDestinationMismatchUrlType
Removing enum PolicyTopicEvidenceDestinationNotWorkingDevice
Removing enum PolicyTopicEvidenceDestinationNotWorkingDnsErrorType
Removing enum PositiveGeoTargetType
Removing enum PreferredContentType
Removing enum PriceExtensionPriceQualifier
Removing enum PriceExtensionPriceUnit
Removing enum PriceExtensionType
Removing enum PricePlaceholderField
Removing enum ProductBiddingCategoryLevel
Removing enum ProductBiddingCategoryStatus
Removing enum ProductChannel
Removing enum ProductChannelExclusivity
Removing enum ProductCondition
Removing enum ProductCustomAttributeIndex
Removing enum ProductTypeLevel
Removing enum PromotionExtensionDiscountModifier
Removing enum PromotionExtensionOccasion
Removing enum PromotionPlaceholderField
Removing enum ProximityRadiusUnits
Removing enum QualityScoreBucket
Removing enum ReachPlanAdLength
Removing enum ReachPlanAgeRange
Removing enum ReachPlanNetwork
Removing enum RealEstatePlaceholderField
Removing enum RecommendationType
Removing enum ResourceChangeOperation
Removing enum ResourceLimitType
Removing enum ResponseContentType
Removing enum SearchEngineResultsPageType
Removing enum SearchTermMatchType
Removing enum SearchTermTargetingStatus
Removing enum SeasonalityEventScope
Removing enum SeasonalityEventStatus
Removing enum ServedAssetFieldType
Removing enum SharedSetStatus
Removing enum SharedSetType
Removing enum SimulationModificationMethod
Removing enum SimulationType
Removing enum SitelinkPlaceholderField
Removing enum Slot
Removing enum SpendingLimitType
Removing enum StructuredSnippetPlaceholderField
Removing enum SummaryRowSetting
Removing enum SystemManagedEntitySource
Removing enum TargetCpaOptInRecommendationGoal
Removing enum TargetImpressionShareLocation
Removing enum TargetingDimension
Removing enum TimeType
Removing enum TravelPlaceholderField
Removing enum UserIdentifierSource
Removing enum UserInterestTaxonomyType
Removing enum UserListAccessStatus
Removing enum UserListClosingReason
Removing enum UserListCombinedRuleOperator
Removing enum UserListCrmDataSourceType
Removing enum UserListDateRuleItemOperator
Removing enum UserListLogicalRuleOperator
Removing enum UserListMembershipStatus
Removing enum UserListNumberRuleItemOperator
Removing enum UserListPrepopulationStatus
Removing enum UserListRuleType
Removing enum UserListSizeRange
Removing enum UserListStringRuleItemOperator
Removing enum UserListType
Removing enum ValueRuleDeviceType
Removing enum ValueRuleGeoLocationMatchType
Removing enum ValueRuleOperation
Removing enum ValueRuleSetAttachmentType
Removing enum ValueRuleSetDimension
Removing enum VanityPharmaDisplayUrlMode
Removing enum VanityPharmaText
Removing enum VideoThumbnail
Removing enum WebpageConditionOperand
Removing enum WebpageConditionOperator
</details>

mikkamp committed:
Correct function comment

mikkamp committed:
Add null type to docblock
